[4.0] Code Generator sometimes hangs and does not work with Benchmark.NET

I have the impression that sometimes the code generator just hangs. I have made this experience in the following branch: https://github.com/OrleansContrib/OrleansDashboard/pull/351

The build just stops and takes forever (or until I stop it). I do not have this issue with other (much larger) projects. Only in the Dashboard project.

Furthermore I cannot get Benchmark.NET to run, it also fails to build.

I'm also getting error when running Benchmark.Net with Microsoft.Orleans.Server 7.1.0

Standard error:
 MSBuild version 17.4.0+18d5aef85 for .NET
C:\code\Asos.Sf.SampleApp\Asos.Sf.SampleApp.Benchmarks\bin\Release\net7.0\85cad078-2b6b-43f5-86ca-0859a3d311df\Orleans.CodeGenerator\Orleans.CodeGenerator.OrleansSerializationSourceGenerator\85cad078-2b6b-43f5-86ca-0859a3d311df.orleans.g.cs(9,93): error CS0118: 'OrleansCodeGen' is a namespace but is used like a type [C:\code\Asos.Sf.SampleApp\Asos.Sf.SampleApp.Benchmarks\bin\Release\net7.0\85cad078-2b6b-43f5-86ca-0859a3d311df\BenchmarkDotNet.Autogenerated.csproj]
C:\code\Asos.Sf.SampleApp\Asos.Sf.SampleApp.Benchmarks\bin\Release\net7.0\85cad078-2b6b-43f5-86ca-0859a3d311df\Orleans.CodeGenerator\Orleans.CodeGenerator.OrleansSerializationSourceGenerator\85cad078-2b6b-43f5-86ca-0859a3d311df.orleans.g.cs(9,110): error CS0103: The name 'cad0782b6b43f586ca0859a3d311df' does not exist in the current context [C:\code\Asos.Sf.SampleApp\Asos.Sf.SampleApp.Benchmarks\bin\Release\net7.0\85cad078-2b6b-43f5-86ca-0859a3d311df\BenchmarkDotNet.Autogenerated.csproj]
C:\code\Asos.Sf.SampleApp\Asos.Sf.SampleApp.Benchmarks\bin\Release\net7.0\85cad078-2b6b-43f5-86ca-0859a3d311df\Orleans.CodeGenerator\Orleans.CodeGenerator.OrleansSerializationSourceGenerator\85cad078-2b6b-43f5-86ca-0859a3d311df.orleans.g.cs(9,12): error CS1729: 'TypeManifestProviderAttribute' does not contain a constructor that takes 3 arguments [C:\code\Asos.Sf.SampleApp\Asos.Sf.SampleApp.Benchmarks\bin\Release\net7.0\85cad078-2b6b-43f5-86ca-0859a3d311df\BenchmarkDotNet.Autogenerated.csproj]
Build FAILED.
C:\code\Asos.Sf.SampleApp\Asos.Sf.SampleApp.Benchmarks\bin\Release\net7.0\85cad078-2b6b-43f5-86ca-0859a3d311df\Orleans.CodeGenerator\Orleans.CodeGenerator.OrleansSerializationSourceGenerator\85cad078-2b6b-43f5-86ca-0859a3d311df.orleans.g.cs(9,93): error CS0118: 'OrleansCodeGen' is a namespace but is used like a type [C:\code\Asos.Sf.SampleApp\Asos.Sf.SampleApp.Benchmarks\bin\Release\net7.0\85cad078-2b6b-43f5-86ca-0859a3d311df\BenchmarkDotNet.Autogenerated.csproj]
C:\code\Asos.Sf.SampleApp\Asos.Sf.SampleApp.Benchmarks\bin\Release\net7.0\85cad078-2b6b-43f5-86ca-0859a3d311df\Orleans.CodeGenerator\Orleans.CodeGenerator.OrleansSerializationSourceGenerator\85cad078-2b6b-43f5-86ca-0859a3d311df.orleans.g.cs(9,110): error CS0103: The name 'cad0782b6b43f586ca0859a3d311df' does not exist in the current context [C:\code\Asos.Sf.SampleApp\Asos.Sf.SampleApp.Benchmarks\bin\Release\net7.0\85cad078-2b6b-43f5-86ca-0859a3d311df\BenchmarkDotNet.Autogenerated.csproj]
C:\code\Asos.Sf.SampleApp\Asos.Sf.SampleApp.Benchmarks\bin\Release\net7.0\85cad078-2b6b-43f5-86ca-0859a3d311df\Orleans.CodeGenerator\Orleans.CodeGenerator.OrleansSerializationSourceGenerator\85cad078-2b6b-43f5-86ca-0859a3d311df.orleans.g.cs(9,12): error CS1729: 'TypeManifestProviderAttribute' does not contain a constructor that takes 3 arguments [C:\code\Asos.Sf.SampleApp\Asos.Sf.SampleApp.Benchmarks\bin\Release\net7.0\85cad078-2b6b-43f5-86ca-0859a3d311df\BenchmarkDotNet.Autogenerated.csproj]
    0 Warning(s)
    3 Error(s)
Time Elapsed 00:00:01.93

// BenchmarkDotNet has failed to build the auto-generated boilerplate code.
// It can be found in C:\code\Asos.Sf.SampleApp\Asos.Sf.SampleApp.Benchmarks\bin\Release\net7.0\85cad078-2b6b-43f5-86ca-0859a3d311df
// Please follow the troubleshooting guide: https://benchmarkdotnet.org/articles/guides/troubleshooting.html

--- a/src/Orleans.CodeGenerator/SyntaxGeneration/Identifier.cs
+++ b/src/Orleans.CodeGenerator/SyntaxGeneration/Identifier.cs
@@ -116,8 +116,17 @@ namespace Orleans.CodeGenerator.SyntaxGeneration
             }
         }
 
-        private static readonly Regex SanitizeIdentifierRegex = new("[-\\.]", RegexOptions.Compiled);
+        private static readonly Regex SanitizeIdentifierRegex = new("^([0-9]+)|([^0-9a-zA-Z_]+)", RegexOptions.Compiled);
 
-        public static string SanitizeIdentifierName(string input) => SanitizeIdentifierRegex.Replace(input, "");
+        public static string SanitizeIdentifierName(string input) => SanitizeIdentifierRegex.Replace(
+            input,
+            static match => match.Value switch
+            {
+                // Prefix leading digits with an '_' to make them a valid identifier.
+                { Length: > 0 } value when char.IsDigit(value[0]) => $"_{value}",
+
+                // Eliminate all other matches by replacing them with an empty string.
+                _ => ""
+            });
     }
 }

